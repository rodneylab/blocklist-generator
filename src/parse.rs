use ahash::RandomState;
use log::trace;
use nom::{
    branch::alt,
    bytes::complete::{tag, take_while_m_n},
    character::complete::{alphanumeric1, multispace1},
    combinator::{recognize, verify},
    multi::{many0_count, many1_count},
    sequence::{pair, tuple},
    IResult,
};
use std::collections::HashSet;
use url::Host;

fn is_digit(c: char) -> bool {
    c.is_ascii_digit()
}

fn parse_ipv4_octet(input: &str) -> IResult<&str, &str> {
    verify(take_while_m_n(1, 3, is_digit), |val: &str| {
        val.parse::<u8>().is_ok()
    })(input)
}

fn parse_hostname_element(input: &str) -> IResult<&str, &str> {
    verify(
        recognize(many1_count(alt((alphanumeric1, tag("-"), tag("_"))))),
        |val: &str| val.len() <= 63,
    )(input)
}

fn parse_hostname(input: &str) -> IResult<&str, &str> {
    recognize(tuple((
        recognize(many1_count(pair(parse_hostname_element, tag(".")))),
        parse_hostname_element,
    )))(input)
}

fn parse_ipv4_address(input: &str) -> IResult<&str, &str> {
    recognize(pair(
        verify(
            many0_count(pair(parse_ipv4_octet, tag("."))),
            |val: &usize| *val <= 3,
        ),
        parse_ipv4_octet,
    ))(input)
}

fn parse_domainlist_line(input: &str) -> Option<&str> {
    // expect "example.com"
    let Ok((_rest, hostname)) = parse_hostname(input) else {
        return None;
    };
    Some(hostname)
}

fn parse_hostfile_line(input: &str) -> Option<&str> {
    // expect "127.0.0.1 example.com"
    let Ok((_rest, (_ipv4_address, _, hostname))) =
        tuple((parse_ipv4_address, multispace1, parse_hostname))(input)
    else {
        return None;
    };
    Some(hostname)
}

pub fn domainlist(file_body: &str, set: &mut HashSet<Host, RandomState>) {
    for line in file_body.lines() {
        if let Some(value) = parse_domainlist_line(line) {
            if let Ok(host_value) = Host::parse(value) {
                set.insert(host_value);
            } else {
                trace!("Unable to parse hostname in line `{value}`");
            }
        } else if !line.is_empty() && line.trim_start()[0..1] != *"#" {
            trace!("Unable to parse `{line}`");
        }
    }
}

pub fn hostfile(file_body: &str, set: &mut HashSet<Host, RandomState>) {
    for line in file_body.lines() {
        if let Some(value) = parse_hostfile_line(line) {
            if let Ok(host_value) = Host::parse(value) {
                set.insert(host_value);
            } else {
                trace!("Unable to parse hostname in line `{value}`");
            }
        } else if !line.is_empty() && line.trim_start()[0..1] != *"#" {
            trace!("Unable to parse `{line}`");
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use ahash::RandomState;
    use fake::{faker, Fake};
    use proptest::{prop_assert_eq, proptest, strategy::Strategy};
    use url::Host;

    use super::{parse_hostfile_line, parse_hostname, parse_ipv4_address, parse_ipv4_octet};
    use crate::parse::{domainlist, hostfile, parse_domainlist_line};

    #[test]
    fn parse_ip4_octet_parses_valid_ipv4_octet() {
        // arrange
        let input_0: &str = "1";
        let input_1: &str = "12";
        let input_2: &str = "123";
        let input_3: &str = "1234";

        // act
        let result_0 = parse_ipv4_octet(input_0);
        let result_1 = parse_ipv4_octet(input_1);
        let result_2 = parse_ipv4_octet(input_2);
        let result_3 = parse_ipv4_octet(input_3);

        // assert
        assert_eq!(result_0, Ok(("", "1")));
        assert_eq!(result_1, Ok(("", "12")));
        assert_eq!(result_2, Ok(("", "123")));
        assert_eq!(result_3, Ok(("4", "123")));
    }

    proptest! {
    #[test]
    fn parse_ip4_octet_parses_valid_ipv4_octet_proptest(octet in 0u8..) {
        // arrange
        let octet_string = octet.to_string();

        // act
        let result = parse_ipv4_octet(&octet_string);

        // assert
        prop_assert_eq!(result, Ok(("", octet_string.as_str())));
    }
    }

    #[test]
    fn parse_ipv4_octet_fails_to_parse_invalid_ipv4_octet() {
        // arrange
        let input_0: &str = "";
        let input_1: &str = ".";
        let input_2: &str = "256";

        // act
        let result_0 = parse_ipv4_octet(input_0);
        let result_1 = parse_ipv4_octet(input_1);
        let result_2 = parse_ipv4_octet(input_2);

        // assert
        assert_eq!(
            result_0,
            Err(nom::Err::Error(nom::error::Error {
                input: "",
                code: nom::error::ErrorKind::TakeWhileMN
            }))
        );
        assert_eq!(
            result_1,
            Err(nom::Err::Error(nom::error::Error {
                input: ".",
                code: nom::error::ErrorKind::TakeWhileMN
            }))
        );
        assert_eq!(
            result_2,
            Err(nom::Err::Error(nom::error::Error {
                input: "256",
                code: nom::error::ErrorKind::Verify
            }))
        );
    }

    fn arb_ipv4_address() -> impl Strategy<Value = String> {
        (0u8.., 0u8.., 0u8.., 0u8..).prop_map(|(octet_0, octet_1, octet_2, octet_3)| {
            format!("{octet_0}.{octet_1}.{octet_2}.{octet_3}")
        })
    }

    proptest! {
    #[test]
    fn parse_ipv4_address_parses_valid_ipv4_proptest(
        ipv4_address in arb_ipv4_address()){
        // arrange

        // act
        let result = parse_ipv4_address(&ipv4_address);

        // assert
        prop_assert_eq!(result, Ok(("", ipv4_address.as_str())));
    }
    }

    #[test]
    fn parse_ipv4_address_parses_valid_ipv4_address() {
        // arrange
        let input_0: &str = "127.0.0.1"; // DevSkim: ignore DS162092 - use of local host IP is in test
        let input_1: &str = "127";
        let input_2: &str = "127.0.0.1 example.com"; // DevSkim: ignore DS162092 - use of local host IP is in test

        // act
        let result_0 = parse_ipv4_address(input_0);
        let result_1 = parse_ipv4_address(input_1);
        let result_2 = parse_ipv4_address(input_2);

        // assert
        assert_eq!(result_0, Ok(("", "127.0.0.1"))); // DevSkim: ignore DS162092 - use of local host IP is in test
        assert_eq!(result_1, Ok(("", "127")));
        assert_eq!(result_2, Ok((" example.com", "127.0.0.1"))); // DevSkim: ignore DS162092 - use of local host IP is in test
    }

    #[test]
    fn parse_ipv4_address_fails_to_parse_invalid_ipv4_address() {
        // arrange
        let input_0: &str = "127.0.0.1.1"; // DevSkim: ignore DS162092 - use of local host IP is in test

        // act
        let result_0 = parse_ipv4_address(input_0);

        // assert
        assert_eq!(
            result_0,
            Err(nom::Err::Error(nom::error::Error {
                input: "127.0.0.1.1", // DevSkim: ignore DS162092 - use of local host IP is in test
                code: nom::error::ErrorKind::Verify
            }))
        );
    }

    #[test]
    fn parse_hostname_parses_valid_hostnames() {
        // arrange
        let input_0: &str = "example.com";
        let input_1: &str = "sub.example.com";
        let input_2: &str = "sub-domain.example.com";
        let input_3: &str = "sub_domain.example.com";

        // act
        let result_0 = parse_hostname(input_0);
        let result_1 = parse_hostname(input_1);
        let result_2 = parse_hostname(input_2);
        let result_3 = parse_hostname(input_3);

        // assert
        assert_eq!(result_0, Ok(("", "example.com")));
        assert_eq!(result_1, Ok(("", "sub.example.com")));
        assert_eq!(result_2, Ok(("", "sub-domain.example.com")));
        assert_eq!(result_3, Ok(("", "sub_domain.example.com")));
    }

    fn arb_subdomain_name() -> impl Strategy<Value = String> {
        "[a-zA-Z0-9_-]{1,63}"
    }

    fn arb_top_level_domain() -> impl Strategy<Value = String> {
        (0u8..1).prop_map(|_| faker::internet::en::DomainSuffix().fake::<String>())
    }

    /// There is no limit on number of subdomains in the standards, though each subdomain should be no longer than
    /// 63 octets and the entire domain name should be no longer than 255 octets.  This test does
    /// not currently push right up to those limits.
    fn arb_domain_name() -> impl Strategy<Value = String> {
        (
            arb_subdomain_name(),
            arb_subdomain_name(),
            arb_subdomain_name(),
            arb_top_level_domain(),
            1u32..4,
        )
            .prop_map(|(subdomain_0, subdomain_1, subdomain_2, tld, subdomains)| {
                match subdomains {
                    1 => format!("{subdomain_0}.{tld}"),
                    2 => format!("{subdomain_0}.{subdomain_1}.{tld}"),
                    3 => format!("{subdomain_0}.{subdomain_1}.{subdomain_2}.{tld}"),
                    _ => unreachable!("Unexpected subdomain count"),
                }
            })
    }

    proptest! {
         #[test]
         fn parse_hostname_parses_valid_hostnames_proptest(
    hostname in arb_domain_name())
     {
             // arrange
             // act
             let result = parse_hostname(&hostname);

             // assert
             prop_assert_eq!(result, Ok(("", hostname.as_str())));
         }
         }

    #[test]
    fn parse_domainlist_line_successfully_handles_invalid_input() {
        // arrange
        let input_0: &str = "com";
        let input_1: &str = "# some annotation";

        // act
        let result_0 = parse_domainlist_line(input_0);
        let result_1 = parse_domainlist_line(input_1);

        // assert
        assert_eq!(result_0, None);
        assert_eq!(result_1, None);
    }

    #[test]
    fn parse_domainlist_line_successfully_parses_valid_input() {
        // arrange
        let input_0: &str = "example.com";
        let input_1: &str = "example.com # some annotation";

        // act
        let result_0 = parse_domainlist_line(input_0);
        let result_1 = parse_domainlist_line(input_1);

        // assert
        assert_eq!(result_0, Some("example.com"));
        assert_eq!(result_1, Some("example.com"));
    }

    proptest! {
         #[test]
    fn parse_domainlist_line_successfully_parses_valid_input_proptest(
    hostname in arb_domain_name())
     {
             // arrange

             // act
             let result = parse_domainlist_line(&hostname);

             // assert
             prop_assert_eq!(result, Some(hostname.as_str()));
         }
         }

    #[test]
    fn parse_hostfile_line_successfully_parses_valid_input() {
        // arrange
        let input_0: &str = "127.0.0.1 example.com"; // DevSkim: ignore DS162092 - use of local host IP is in test

        // act
        let result_0 = parse_hostfile_line(input_0);

        // assert
        assert_eq!(result_0, Some("example.com"));
    }

    proptest! {
         #[test]
    fn parse_hostfile_line_successfully_parses_valid_input_proptest(
        ipv4_address in arb_ipv4_address(),
    hostname in arb_domain_name())
     {
             // arrange
             let line = format!("{ipv4_address} {hostname}");

             // act
             let result = parse_hostfile_line(&line);

             // assert
             prop_assert_eq!(result, Some(hostname.as_str()));
         }
         }

    #[test]
    fn domainlist_successfully_parses_valid_input() {
        // arrange
        let input = r"example.com
another-example.com # some annotation

# more annotation
subdomain-which-is-too-long-012345679012345678901234567890123456.com
final-example.com";
        let hasher = RandomState::new();
        let mut hash_set: HashSet<Host, RandomState> = HashSet::with_hasher(hasher);

        // act
        domainlist(input, &mut hash_set);

        // assert
        assert_eq!(hash_set.len(), 3);
        assert!(hash_set.contains(&Host::parse("example.com").unwrap()));
        assert!(hash_set.contains(&Host::parse("another-example.com").unwrap()));
        assert!(hash_set.contains(&Host::parse("final-example.com").unwrap()));
    }

    #[test]
    fn hostfile_successfully_parses_valid_input() {
        // arrange
        let _ = env_logger::builder().is_test(true).try_init();
        let input = "127.0.0.1\texample.com\n0.0.0.0 another-example.com # some annotation\n\n# more annotation\n0.0.0.0\t\tsubdomain-which-is-too-long-012345679012345678901234567890123456.com\n0.0.0.0\tfinal-example.com";
        let hasher = RandomState::new();
        let mut hash_set: HashSet<Host, RandomState> = HashSet::with_hasher(hasher);

        // act
        hostfile(input, &mut hash_set);

        // assert
        assert_eq!(hash_set.len(), 3);
        assert!(hash_set.contains(&Host::parse("example.com").unwrap()));
        assert!(hash_set.contains(&Host::parse("another-example.com").unwrap()));
        assert!(hash_set.contains(&Host::parse("final-example.com").unwrap()));
    }
}
