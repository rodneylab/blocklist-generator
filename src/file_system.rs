use std::{
    collections::HashSet,
    fs::{self, File},
    io::Write,
    path::{Path, PathBuf},
};

use ahash::RandomState;
use anyhow::Context;
use askama::Template;
use humansize::{format_size, DECIMAL};
use log::{error, info};
use serde::Deserialize;
use url::Host;

use crate::parse::domainlist as parse_domainlist;

#[derive(Debug, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Blocklists {
    pub hosts_file_blocklist_urls: Vec<String>,
    pub domain_blocklist_urls: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Filters {
    pub allowed_names: Option<Vec<String>>,

    #[cfg_attr(not(test), expect(dead_code))]
    pub blocked_names: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Config {
    pub blocklists: Blocklists,
    pub filters: Option<Filters>,
}

pub fn get_config_from_file<P: AsRef<Path>>(config_file_path: P) -> anyhow::Result<Config> {
    let config_file_content = fs::read_to_string(&config_file_path).with_context(|| {
        format!(
            "Failed to open or read config file `{}`",
            config_file_path.as_ref().display()
        )
    })?;

    match toml::from_str(&config_file_content) {
        Ok(value) => Ok(value),
        Err(_) => anyhow::bail!(
            "Failed to parse config file `{}`.  Check it is valid.",
            config_file_path.as_ref().display()
        ),
    }
}

pub fn get_custom_blocked_names<P: AsRef<Path>>(
    blocked_names_path: P,
    set: &mut HashSet<Host, RandomState>,
) {
    let blocked_names_display_path = blocked_names_path.as_ref().display().to_string();
    let blocked_names_content = if let Ok(value) = fs::read_to_string(blocked_names_path) {
        Some(value)
    } else {
        info!("No custom blocked names file found at `{blocked_names_display_path}.",);
        None
    };
    if let Some(value) = blocked_names_content {
        parse_domainlist(&value, set);
    };
}

#[derive(Template)]
#[template(escape = "none", path = "blocklist.rpz")]
struct BlocklistRPZTemplate<'a> {
    domains: &'a str,
}

fn domain_to_blocklist_rpz_domain(host: &Host) -> String {
    let domain = host.to_string();
    format!("{domain}\tCNAME\t.\n*.{domain}\tCNAME\t.\n")
}

fn domain_to_unbound_local_zone(host: &Host) -> String {
    let domain = host.to_string();
    format!("local-zone: \"{domain}\" always_nxdomain\n")
}

fn write_to_file<P: AsRef<Path>>(content: &str, output_path: &P) {
    let output_display_path = output_path.as_ref().display().to_string();
    let Ok(mut outfile) = File::create(output_path) else {
        error!("Unable to create output file");
        panic!("Error creating output file {output_display_path}")
    };
    if outfile.write_all(content.as_bytes()).is_err() {
        error!("Unable to write to output file {output_display_path}");
        panic!("Error writing to output file");
    }
    info!("Wrote data to file: {output_display_path}");
}

fn print_output_file_metadata<P: AsRef<Path>>(output_path: &P) {
    if let Ok(value) = fs::metadata(output_path) {
        let bytes = value.len();
        let display_bytes = format_size(bytes, DECIMAL);
        let display_path = output_path.as_ref().display();
        std::println!("Written {display_bytes} to {display_path}");
    }
}

pub fn write_domain_blocklist_file(blocklist_domains: &[Host]) {
    let domains = blocklist_domains
        .iter()
        .fold(String::new(), |mut acc, val| {
            acc.push_str(&val.to_string());
            acc.push('\n');
            acc
        });
    let output_path = PathBuf::from("./domain-blocklist.txt");
    write_to_file(&domains, &output_path);
    print_output_file_metadata(&output_path);
}

pub fn write_blocklist_rpz_file(blocklist_domains: &[Host]) {
    let domains = blocklist_domains
        .iter()
        .fold(String::new(), |mut acc, val| {
            acc.push_str(&domain_to_blocklist_rpz_domain(val));
            acc
        });
    let template = BlocklistRPZTemplate { domains: &domains };
    let file_content = template
        .render()
        .expect("Unexpected error rendering template");
    let output_path = PathBuf::from("./blocklist.rpz");
    write_to_file(&file_content, &output_path);
    print_output_file_metadata(&output_path);
}

pub fn write_unbound_local_zone_file(blocklist_domains: &[Host]) {
    let output_path = PathBuf::from("./zone-block-general.conf");
    let file_content = blocklist_domains
        .iter()
        .fold(String::new(), |mut acc, val| {
            acc.push_str(&domain_to_unbound_local_zone(val));
            acc
        });
    write_to_file(&file_content, &output_path);
    print_output_file_metadata(&output_path);
}

#[cfg(test)]
mod tests {
    use assert_fs::fixture::{FileWriteStr, PathChild};

    use crate::file_system::get_config_from_file;

    #[test]
    fn get_config_from_file_successfully_parses_valid_file() {
        let config_content = r#"[blocklists]
hosts_file_blocklist_urls = [
  "https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts;showintro=0",
  "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling-porn/hosts",
  "https://raw.githubusercontent.com/anudeepND/blacklist/master/adservers.txt",
  "https://reddestdream.github.io/Projects/MinimalHosts/etc/MinimalHostsBlocker/minimalhosts",
]
domain_blocklist_urls = [
  "https://quidsup.net/notrack/blocklist.php?download=annoyancedomains",
  "https://quidsup.net/notrack/blocklist.php?download=malwaredomains",
  "https://quidsup.net/notrack/blocklist.php?download=trackersdomains",
  "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt",
  "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt",
  "https://v.firebog.net/hosts/AdguardDNS.txt",
  "https://v.firebog.net/hosts/Easyprivacy.txt",
]

[filters]
allowed_names = [
  "0.0.0.0",
  "127.0.0.1",
  "255.255.255.255",
  "example.com",
  "another.example.com",
]
"#;
        let temp_dir = assert_fs::TempDir::new().unwrap();
        let _ = temp_dir
            .child("blocklist-generator.toml")
            .write_str(config_content);
        let config_path = temp_dir.join("blocklist-generator.toml");

        // act
        let outcome = get_config_from_file(config_path).unwrap();

        // assert
        insta::assert_json_snapshot!(outcome);
    }

    #[test]
    fn get_config_from_file_returns_error_on_config_file_system_error() {
        let temp_dir = assert_fs::TempDir::new().unwrap();
        // file not created deliberately
        let config_path = temp_dir.join("does-not-exist-generator.toml");

        // act
        let outcome = get_config_from_file(&config_path).unwrap_err();

        // assert
        assert_eq!(
            format!("{outcome}"),
            format!(
                "Failed to open or read config file `{}`",
                &config_path.display()
            )
        );
        let mut chain = outcome.chain();
        assert_eq!(
            chain.next().map(|val| format!("{val}")),
            Some(format!(
                "Failed to open or read config file `{}`",
                &config_path.display()
            ))
        );
        assert_eq!(
            chain.next().map(|val| format!("{val}")),
            Some(String::from("No such file or directory (os error 2)"))
        );
        assert!(chain.next().is_none());
    }

    #[test]
    fn get_config_from_file_returns_error_on_invalid_config_file() {
        let config_content = r#"[blocklists]
hosts_file_blocklist_urls = [
  "https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts;showintro=0",
  "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling-porn/hosts",
  "https://raw.githubusercontent.com/anudeepND/blacklist/master/adservers.txt",
  "https://reddestdream.github.io/Projects/MinimalHosts/etc/MinimalHostsBlocker/minimalhosts",
]
domain_blocklist_urls = [
  "https://quidsup.net/notrack/blocklist.php?download=annoyancedomains",
  "https://quidsup.net/notrack/blocklist.php?download=malwaredomains",
  "https://quidsup.net/notrack/blocklist.php?download=trackersdomains",
  "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt",
  "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt",
  "https://v.firebog.net/hosts/AdguardDNS.txt",
  "https://v.firebog.net/hosts/Easyprivacy.txt",
]

[filter"#;
        let temp_dir = assert_fs::TempDir::new().unwrap();
        let _ = temp_dir
            .child("blocklist-generator.toml")
            .write_str(config_content);
        let config_path = temp_dir.join("blocklist-generator.toml");

        // act
        let outcome = get_config_from_file(&config_path).unwrap_err();

        // assert
        assert_eq!(
            format!("{outcome}"),
            format!(
                "Failed to parse config file `{}`.  Check it is valid.",
                &config_path.display()
            )
        );
        let mut chain = outcome.chain();
        assert_eq!(
            chain.next().map(|val| format!("{val}")),
            Some(format!(
                "Failed to parse config file `{}`.  Check it is valid.",
                &config_path.display()
            ))
        );
        assert!(chain.next().is_none());
    }
}
