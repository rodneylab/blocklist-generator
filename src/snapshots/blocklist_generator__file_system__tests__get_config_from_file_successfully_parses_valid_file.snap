---
source: src/file_system.rs
expression: outcome
snapshot_kind: text
---
{
  "blocklists": {
    "hosts_file_blocklist_urls": [
      "https://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts;showintro=0",
      "https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling-porn/hosts",
      "https://raw.githubusercontent.com/anudeepND/blacklist/master/adservers.txt",
      "https://reddestdream.github.io/Projects/MinimalHosts/etc/MinimalHostsBlocker/minimalhosts"
    ],
    "domain_blocklist_urls": [
      "https://quidsup.net/notrack/blocklist.php?download=annoyancedomains",
      "https://quidsup.net/notrack/blocklist.php?download=malwaredomains",
      "https://quidsup.net/notrack/blocklist.php?download=trackersdomains",
      "https://s3.amazonaws.com/lists.disconnect.me/simple_ad.txt",
      "https://s3.amazonaws.com/lists.disconnect.me/simple_tracking.txt",
      "https://v.firebog.net/hosts/AdguardDNS.txt",
      "https://v.firebog.net/hosts/Easyprivacy.txt"
    ]
  },
  "filters": {
    "allowed_names": [
      "0.0.0.0",
      "127.0.0.1",
      "255.255.255.255",
      "example.com",
      "another.example.com"
    ],
    "blocked_names": null
  }
}
